x/stakeibc/keeper/callbacks.go

```diff
@@ -156,3 +160,113 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+			if qNumTokens.Uint64() < v.DelegationAmt {
+				// TODO(later) add some safety checks here (e.g. we could query the slashing module to confirm the decr in tokens was due to slash)
+				// update our records of the total stakedbal and of the validator's delegation amt
+				// NOTE:  we assume any decrease in delegation amt that's not tracked via records is a slash
+				slashAmt := v.DelegationAmt - qNumTokens.Uint64()
+				slashPct := sdk.NewDec(cast.ToInt64(slashAmt)).Quo(sdk.NewDec(cast.ToInt64(v.DelegationAmt)))
+				k.Logger(ctx).Info(fmt.Sprintf("ICQ'd delAmt mismatch zone %s validator %s delegator %s records was %d icq shows %d slashAmt %d slashPct %d... UPDATING!",
+					zone.ChainId, v.Address, qdel.DelegatorAddress, v.DelegationAmt, qNumTokens, slashAmt, slashPct))
+				k.Logger(ctx).Info(fmt.Sprintf("ICQ'd: slashPct: %v", slashPct))
+				// TODO (not priority) move rate limiting logic to new rate limiting module
+				if slashPct.GT(sdk.NewDec(10).Quo(sdk.NewDec(100))) {
+					k.Logger(ctx).Info(fmt.Sprintf("DELCB | slashed but ABORTING bc slash GT10pct: query shows slash of %v", slashPct))
+					return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "DELCB | slashed but ABORTING bc slash GT10pct: query shows slash of %v", slashPct)
+				}
+				// slash the validator's weight
+				weightMul := sdk.NewDec(qNumTokens.Int64()).Quo(sdk.NewDec(cast.ToInt64(v.DelegationAmt)))
+
+				zone.StakedBal -= cast.ToInt64(slashAmt)
+				v.DelegationAmt -= slashAmt
+				v.Weight = sdk.NewDec(cast.ToInt64(v.Weight)).Mul(weightMul).TruncateInt().Uint64()
+			}
+			// write back to state and break (reset TokensFromShares for clarity, so we're not tempted to use it again later)
+			v.TokensFromShares = sdk.NewDec(0)
+			zone.Validators[i] = v
+			k.SetHostZone(ctx, zone)
+			break
+		}
+	}
+	return nil
+}
+
+// ValidatorCallback is a callback handler for validator queries.
+func ValidatorCallback(k Keeper, ctx sdk.Context, args []byte, query icqtypes.Query) error {
+	zone, found := k.GetHostZone(ctx, query.GetChainId())
+	if !found {
+		return fmt.Errorf("no registered zone for chain id: %s", query.GetChainId())
+	}
+	qval := stakingtypes.Validator{}
+	err := k.cdc.Unmarshal(args, &qval)
+	if err != nil {
+		k.Logger(ctx).Error("unable to unmarshal qval info for zone", "zone", zone.ChainId, "err", err)
+		return err
+	}
+	k.Logger(ctx).Info(fmt.Sprintf("ValidatorCallback: zone %v qval %v", zone.ChainId, qval))
+
+	// ensure ICQ can be issued now! else fail the callback
+	valid, err := k.IsWithinBufferWindow(ctx)
+	if err != nil {
+		return err
+	} else if !valid {
+		return nil
+	}
+
+	// set the validator's conversion rate
+	for i, v := range zone.Validators {
+		if v.Address == qval.OperatorAddress {
+			// converting 1.0 gives us the exchange rate to later use in the next CB
+			v.TokensFromShares = qval.TokensFromShares(sdk.NewDec(1.0))
```

Comment:
can you elaborate on why this is 1.0?

Reply:
`TokensFromShares` is a fn that multiplies its input by a constant (which is derived within the `TokensFromShares ` fn itself). To get that constant out of the fn, we can pass in `1.0`. I prefer this approach to copying over the internal logic of `TokensFromShares`, because with this approach, even if the internals of `TokensFromShares`, change, we're safe (bc we use the cosmos sdk core module version)